### ISSUE TITLE
By default initialize all variables and use a bit longer syntax for uninitialized/unformed  variables. 

### ISSUE THREAD

**Issue by libbooze** [carbon-language/carbon-lang#1479]
Currently p0257 proposes this to mean uninitialized int:
`var unformed_x: Int;`

I think much more sensible default is

```
// initialized to 0
var  x: Int;
```
```
// initialized to garbage
var  x: Int{?};
```

Not sure about `{?}` I just wanted a short way to say unknown, maybe people can think of a better syntax.

**Comment by emillaine** [comment]
Zig uses [`undefined`](https://ziglang.org/documentation/master/#undefined) for this:

```
var x: Int = undefined;
```

**Comment by chandlerc** [comment]
This was discussed, but it is somewhat complex.

As the proposal discusses, it isn't clear there *is* a sensible default in many cases. We have a specific concern with initializing variables to non-sensible "default" values and being unable to detect bugs as a consequence. So there was a strong desire to not have `var x: Int;` be indistinguishable from a bug and an intended initialization.

As for simply using a longer spelling like `undefined`, there was no strong desire at the time the proposal was discussed once we had an unformed state and the ability to track and catch errors reliably. Personally, the above rationale still applies for me -- I don't want to overly incentivise meaningless initialization to avoid typing more.

The security and other risks of course are real, which is why we worked hard to have the concept of an unformed state that ensures we can prevent security-risking UB in these cases even in the face of bugs.

I'll mark this as a question for the leads as to whether we should explore this more since it isn't *exactly* the same as the already discussed alternatives in #257.

**Comment by libbooze** [comment]
> As the proposal discusses, it isn't clear there _is_ a sensible default in many cases. We have a specific concern with initializing variables to non-sensible "default" values and being unable to detect bugs as a consequence. So there was a strong desire to not have `var x: Int;` be indistinguishable from a bug and an intended initialization.

I had similar discussion with an colleague few years ago when cleaning up warnings in an old C++ codebase (initializing everything to 0 will disable sanitizer detection) so I understand what you are saying.

Still this choice is optimizing for a rare case when people actually want to leave variable uninitialized, is contrary to what most(all?) modern languages use. Not to mention that our good old friend C++ will sometimes initialize int to 0 (static vars, value of a `map<string,int>` when you use `[]` with nonexisting key) so simplifying it will make it easier to learn the language...

Proper way to handle this would be to use an optional or a tag that guarantees variable is checked for initialization before use(either at compile time or runtime).

so something like

```
var x: Int; // 0
var x: Int{?} // garbage
var x Int{!?} // garbage, runtime check before use, byte or bit of storage overhead
var x Int{...} // compiler check that variable will be initialized before use, no runtime overhead, just compile time overhead, obviously can not always work
```

So I understand your motivation but I really hope Carbon team reconsiders this.

**Comment by piotrdz** [comment]
I also second this proposal.

This has always been a source of low-level frustration for me in C and C++, as the language does not make it very clear when a variable starts life with uninitialized data. In fact, you have to look for a form of initialization, or a language rule that enforces it, to prove that a variable is initialized.

I think that having an explicit syntax to flag uninitialized data would be a great help, especially thinking about newcomers to the language who may be used to initialization conventions from other languages.

**Comment by zygoloid** [comment]
I find that treating an omitted initializer as meaning "pick a specific default value" is error-prone and optimizes for the writer at the expense of the reader and potentially at the expense of correctness, so I don't think we should make that change -- while it's not explicitly written up in #257, this is something we discussed and I don't think we have any new information in this issue regarding that question.

I wouldn't mind having some kind of explicit syntax to say "I meant to leave this uninitialized; I'll initialize it later" as a cue to a reader. I think this would be OK:

```
var n: i32; // error, must either give an initializer or say this is uninitialized
var n: i32{?}; // OK, uninitialized (however this is spelled)
var n: i32 = 0; // OK
```

If someone wants to make the case that explicit is better than implicit here, and consider some set of syntactic choices, I think we should be open to that.

**Comment by chandlerc** [comment]
> If someone wants to make the case that explicit is better than implicit here, and consider some set of syntactic choices, I think we should be open to that.

FWIW, I agree with this.

However, I think that case would be much more effectively made with some experience using the current approach. This seems especially easy to introduce with a tool even.

So I'd suggest we say "not right now", and revisit this when we have more significant amounts of Carbon code and can start to see how the experience plays out with users given the more rich semantic model that Carbon provides here.

One thing I want to note: this is purely a *syntactic* question. Carbon's design specifically provides a robust mechanism to harden "uninitialized" values against bugs and undefined behavior. That capability is essential IMO.

**Comment by libbooze** [comment]
> I find that treating an omitted initializer as meaning "pick a specific default value" is error-prone and optimizes for the writer at the expense of the reader and potentially at the expense of correctnes

This is a long and complex discussion that comes up in many places. When explicit code becomes spammy enough that it is actually hurting readability? You are probably familiar in your work/code review with auto wars. :) 
C++ aside a lot of people(including me) dislike that go is(was? - not following for years) obnoxiously explicit with error handling, but some people like it since it is easy for people to learn it and hard to miss it. On the other hand I think that deducing this in C++23 is not worth the benefits, since it makes trivial stuff like member fns hard to understand for beginners(assuming they join a team where the feature is used). And many similar examples...

tl;dr optimizing for the writer does not always mean that reader will suffer, but I see your point.

 

> One thing I want to note: this is purely a _syntactic_ question. Carbon's design specifically provides a robust mechanism to harden "uninitialized" values against bugs and undefined behavior. That capability is essential IMO.

Not sure what is the current state of language(and godbolt seems dead atm) but big help would be if we could to conditional initialization without cute but ugly C++ hacks like IIFE.
for example Rust is really nice in this regard:
https://stackoverflow.com/a/53981839/700825

**Comment by chandlerc** [comment]
After discussing, the leads decided "not right now". We can easily revisit this when we have (significantly) more experience with the current approach.
